core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -817,4 +830,65 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
             }
         return defaultValue;
     }
+
+    private static void fireOnFailure(final NodeProvisioner.PlannedNode plannedNode, final Throwable cause) {
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onFailure(plannedNode, cause);
+            } catch (Throwable e) {
+                LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while "
+                        + "processing onFailure() listener call in " + cl + " for agent "
+                        + plannedNode.displayName, e);
+            }
+        }
+    }
+
+    private static void fireOnRollback(final NodeProvisioner.PlannedNode plannedNode, final Throwable cause) {
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onRollback(plannedNode, cause);
+            } catch (Throwable e) {
+                LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while "
+                        + "processing onRollback() listener call in " + cl + " for agent "
+                        + plannedNode.displayName, e);
+            }
+        }
+    }
+
+    private static void fireOnComplete(final NodeProvisioner.PlannedNode plannedNode, final Node newNode) {
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onComplete(plannedNode, newNode);
+            } catch (Throwable e) {
+                LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while "
+                        + "processing onComplete() listener call in " + cl + " for agent "
+                        + plannedNode.displayName, e);
+            }
+        }
+    }
+
+    private static void fireOnCommit(final NodeProvisioner.PlannedNode plannedNode, final Node newNode) {
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onCommit(plannedNode, newNode);
+            } catch (Throwable e) {
```

Comment:
Throwable suppresses Errors like OutOfMemory exception. These errors are supposed to be fatal in Java.
I would rather handle Exceptions only (FindBugs does not like it IIRC) or propagate Errors